enhancements/cvo/clusteroperator-resource-handling.md

```diff
@@ -0,0 +1,100 @@
+---
+title: clusteroperator-resource-handling
+authors:
+  - "@deads2k"
+reviewers:
+  - "@abhinavdahiya"
+  - "@sdodson"
+approvers:
+  - "@abhinavdahiya"
+  - "@derekwaynecarr"
+creation-date: 2019-12-11
+last-updated: 2019-12-11
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# ClusterOperator Resource Handling
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+`clusteroperators.config.openshift.io` are used to determine success of installation.
+They are also used to drive collection of debugging data from tools like `oc adm inspect` and `oc adm must-gather`.
+The `clusteroperator` resources in a payload should always be present, even before installation of particular
+operators to see which `clusteroperators` need to report in and to allow establishing `.status.relatedResources` before
+an operator pod runs.
+This is critical for debugging clusters that fail to install or fail to upgrade with new operators present.
+
+## Motivation
+
+Debugging information for clusters that succeed in bootstrapping, but fail during installation is missing most
+of the data required to resolve the issue via must-gather.
+
+### Goals
+
+1. allow collection of debugging data for failed installs using normal tools. 
+
+### Non-Goals
+
+1. create a new tool to gather data for failed installs after bootstrapping.
+
+## Proposal
+
+1. `clusteroperator` resources in the payload should be created with the required status conditions (available, progressing,
+   degraded) set to `Unknown`.
+2. `clusteroperator` creation by the CVO needs to honor or update `.status.relatedResources`.  This requires updating
+    status after the creation.
+3.  `clusteroperator` resources in the payload should all be created immediately regardless of where in the payload ordering
+    they are located.  This ensures that they are always present during collection.
+4.  The CVO waiting logic on `clusteroperator` remains the same.
+
```

Comment:
if a user deleted the clusteroperator object, is the responsibilty of CVO to create and set these defaults again?

Reply:
> if a user deleted the clusteroperator object, is the responsibilty of CVO to create and set these defaults again?

it would be a race with clusteroperators.  I don't think the distinction matters.

Reply:
> it would be a race with clusteroperators.

Racing in general is not great, but it's mitigated in the ClusterOperator case by having no `spec` and very little `metadata` (e.g. [here][1]).

[1]: https://storage.googleapis.com/origin-ci-test/logs/release-openshift-ocp-installer-e2e-aws-4.4/182/artifacts/e2e-aws/must-gather/quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-5340001b4de0afdf9d37598e91117ef7fcab309b522f526127d9b2ba1c48ecd8/cluster-scoped-resources/config.openshift.io/clusteroperators/authentication.yaml